src/shared/clients/devenvClient.ts

```diff
@@ -103,6 +104,123 @@ export class DevEnvClient implements vscode.Disposable {
         // `Authorization` _should_ have two parameters (RFC 7235), MDE should probably fix that
         headers: { Authorization: this.authToken },
     })
+
+    /**
+     * WARNING: You should use {@link DevEnvHeartbeat} unless you have a reason not to.
+     */
+    async updateActivity(timestamp: number = Date.now()): Promise<number> {
+        await this.got.put('activity', { json: { timestamp: timestamp.toString() } })
+        return timestamp
+    }
+
+    /**
+     * WARNING: You should use {@link DevEnvHeartbeat} unless you have a reason not to.
+     */
+    async getActivity(): Promise<number | undefined> {
+        const response = await this.got<GetActivityResponse>('activity')
+        return response.body.timestamp ? parseInt(response.body.timestamp) : undefined
+    }
+}
+
+/**
+ * This allows you to easily work with Dev Env heartbeats (timestamps).
+ * 
+ * In this context, a 'heartbeat' is a timestamp that the server uses to
+ * determine when the user was last active.
+ */
+export class DevEnvHeartbeat implements vscode.Disposable{
+    private heartbeatEventEmitter = new vscode.EventEmitter<number>()
+    private ideActivityListener: vscode.Disposable | undefined
+    /** The last known heartbeat timestamp, but there could be a newer one on the server. */
+    private lastLocalHeartbeat: number | undefined
+    private extensionUserActivity: ExtensionUserActivity
+
+    static readonly heartbeatDelay = 10_000
+
+    /**
+     * Returns an instance if the heartbeat mechanism is confirmed to be working.
+     */
+    static async instanceIfHeartbeatWorking(client: DevEnvClient, extensionUserActivity?: ExtensionUserActivity): Promise<DevEnvHeartbeat | undefined> {
+        try {
+            await client.getActivity()
+        }
+        catch (e) {
+            const error = e instanceof HTTPError ? e.response.body : e
+            getLogger().error(`DevEnvHeartbeat: Heartbeat API failed:%s`, error)
+            return undefined
+        }
+        
+        return new DevEnvHeartbeat(client, extensionUserActivity)
+    }
+
+    private constructor(private readonly client: DevEnvClient, extensionUserActivity?: ExtensionUserActivity) {
+        this.extensionUserActivity = extensionUserActivity ?? new ExtensionUserActivity(DevEnvHeartbeat.heartbeatDelay)
+    }
+
+    /** Send heartbeat timestamp to the Dev Env */
+    async sendHeartbeat(timestamp: number = Date.now()): Promise<number> {
+        await this.client.updateActivity()
+        this.lastLocalHeartbeat = timestamp
+        this.heartbeatEventEmitter.fire(timestamp)
+        return timestamp
+    }
+
+    /** Get the latest heartbeat timestamp from the Dev Env */
+    async getLatestHeartbeat(): Promise<number | undefined> {
```

Comment:
have the "activity" concept, so introducing another name for it adds indirection but doesn't gain much. Can we avoid adding "heartbeat" as a quasi-synonymous concept?

Reply:
I like using the "heartbeat" keyword, but the api is called "activity". Should I switch everything to just match the API (use "activity")?

Reply:
If we can avoid extra jargon we should. If heartbeat is a necessary _new_ concept in addition to "activity" then it may be justified.